### PR TITLE
fix protocol classification flakiness

### PR DESCRIPTION
**Pull request by guyarb**
### What does this PR do?

Fixed flakiness of protocol classification tests

We have found race condition that can happen as socket filters run asynchronous to each other, and one socket filter classifies the protocol (for example gRPC) while the second isn't able to classify (not the first packet of the protocol, or the response is not classifiable) and they are overriding each other.

The fix is to move the classification code to happen for every packet, (the heaviest part of the code) and to pull the cached protocol from the map closer to its update.
 

### Motivation

Flaky tests in the CI 

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/network/tracer/tracer_classification_test.go [hunk 1]
```diff
@@ -246,33 +246,50 @@ func testProtocolClassification(t *testing.T, cfg *config.Config, clientHost, ta
...
 				break
 			}
 		}
 
-		return foundOutgoingWithProtocol && foundIncomingWithProtocol
-	}, 3*time.Second, 500*time.Millisecond, "couldn't find incoming and outgoing connections with protocol %d for server address %s and target address %s", expectedProtocol, serverAddr, targetAddr)
+		return bool(foundOutgoingWithProtocol) && bool(foundIncomingWithProtocol)
+	}, 5*time.Second, 500*time.Millisecond, "couldn't find incoming and outgoing connections with protocol %d for "+
+		"server address %s and target address %s.\nIncoming: %v\nOutgoing: %v\nfound incoming with protocol: %s\nfound "+
+		"outgoing with protocol: %s", expectedProtocol, serverAddr, targetAddr, &incomingConns, &outgoingConns, &foundIncomingWithProtocol, &foundOutgoingWithProtocol)
```
Comment (brycekahle) [inline pkg/network/tracer/tracer_classification_test.go:294]: Why not use `%v` here for the bools?

Comment (guyarb) [reply pkg/network/tracer/tracer_classification_test.go:294]: %v will print the original value as it is being passed by value to require.Eventually

```func TestMe(t *testing.T) {
	val := false
	require.Eventuallyf(t, func() bool {
		val = true
		return false
	}, time.Second, 10*time.Millisecond, "val is %v", val)
}
```

Comment (brycekahle) [reply pkg/network/tracer/tracer_classification_test.go:294]: Ah, I understand now. There is probably a better way to handle this, but I suppose is fine for now.

File: pkg/network/ebpf/c/protocol-classification-helpers.h [hunk 2]
```diff
@@ -238,6 +238,13 @@ static __always_inline void protocol_classifier_entrypoint(struct __sk_buff *skb
         return;
     }
 
+    protocol_t classified_protocol = PROTOCOL_UNCLASSIFIED;
+    char request_fragment[CLASSIFICATION_MAX_BUFFER];
+    bpf_memset(request_fragment, 0, sizeof(request_fragment));
+    read_into_buffer_for_classification((char *)request_fragment, skb, &skb_info);
```
Comment (brycekahle) [inline pkg/network/ebpf/c/protocol-classification-helpers.h:244]: Does this mean you will try to classify all buffers now? The removed comment seemed to indicate the `if` was guarding against doing that.

**Review by richl-dd — APPROVED** [review]

File: pkg/network/tracer/connection/kprobe/tracer.go [hunk 3]
```diff
@@ -26,7 +26,6 @@ import (
 	"github.com/DataDog/datadog-agent/pkg/network/config"
```
Comment (Influ20) [inline pkg/network/tracer/connection/kprobe/tracer.go:26]: Jd

File: pkg/network/tracer/connection/kprobe/tracer.go [hunk 4]
```diff
@@ -26,7 +26,6 @@ import (
...
 	netebpf "github.com/DataDog/datadog-agent/pkg/network/ebpf"
 	"github.com/DataDog/datadog-agent/pkg/network/ebpf/probes"
```
Comment (Influ20) [inline pkg/network/tracer/connection/kprobe/tracer.go:28]: D

**Review by Influ20 — APPROVED** [review]

**Review by Influ20 — APPROVED** [review]
Yra

**Pull request closed without merging**